Modbus Proxy leitet Funktionstyp "0x02 Read Discrete Inputs" nicht weiter

### Describe the bug

Über den Modbus Proxy lassen sich keine Anfragen vom Modbus Funktionstyp "02 (0x02) Read Discrete Inputs" durch schicken. Als Antwort kommt nur eine Modbus Fehlermeldung zurück. Der Modbus Proxy schreibt bei einer solchen Anfrage auch nichts ins Log.

### Steps to reproduce

1. Modbus Proxy (ohne readonly) konfigurieren, Ziel bei mir: eine E3DC/Wallbe/Phoenix Wallbox
2. z.B. mit Modbus Master eine Anfrage vom Typ "Read Discrete" auf die Register 200 bis 208 an den evcc Modbus Proxy schicken (00 3a 00 00 00 06 ff 02 00 c8 00 09)
3. Modbus Proxy loggt nichts (!)
4. Als Antwort an Modbus Master kommt Fehler zurück (00 3a 00 00 00 03 ff 82 01)
5. Test ob Modbus Proxy und Log funktioniert: mit Modbus Master eine Anfrage "Read Coils" auf Register 480 schicken
6. Anfrage und Antwort steht im evcc Log

Rede ich direkt mit der Wallbox, klappt die Anfrage:
TX: 00 3e 00 00 00 06 ff 02 00 c8 00 09 
RX: 00 3e 00 00 00 05 ff 02 02 a6 01 

### Configuration details

```yaml
modbusproxy:
- port: 502
  uri: 192.168.1.125:502
#  readonly: true
```


### Log details

```text
Log TRACE:

sudo journalctl -fau evcc | grep proxy
Okt 31 14:48:03 fhem-heizung evcc[22593]: [proxy-502] DEBUG 2023/10/31 14:48:03 modbus proxy for 192.168.1.125:502 listening at :502
Okt 31 14:49:27 fhem-heizung evcc[22593]: [proxy-502] TRACE 2023/10/31 14:49:27 read coil: id 255 addr 480 qty 1
Okt 31 14:49:27 fhem-heizung evcc[22593]: [proxy-502] TRACE 2023/10/31 14:49:27 read coil: 00
```


### What type of operating system are you running?

Linux

### Version

0.121.5

Im Readme finde ich nur Infos, dass ich go und node.js  brauche. Habe ich auch instaliert und mir den Quellcode heruntergeladen. Wie lautet der go build Befehl, damit er mir eine exe aus meinem lokalen Repository baut?

Rufe ich das hier aus dem Quellcodeverzeichnis heraus auf
`go build -o d:\Sourcecode\evcc\evcc_proxytest.exe`
so lädt er einen Haufen Dateien herunter, endet dann aber mit
`main.go:16:13: pattern dist: no matching files found`

--- a/server/modbus/handler.go
+++ b/server/modbus/handler.go
@@ -72,7 +72,7 @@ func coilsToBytes(b []bool) []byte {
 	return res
 }
 
-func (h *handler) coilsToResult(op string, qty uint16, b []byte, err error) ([]bool, error) {
+func (h *handler) bytesToBoolResult(op string, qty uint16, b []byte, err error) ([]bool, error) {
 	h.logResult(op, b, err)
 
 	var modbusError *gridx.Error
@@ -96,6 +96,12 @@ LOOP:
 	return res, err
 }
 
+func (h *handler) HandleDiscreteInputs(req *mbserver.DiscreteInputsRequest) ([]bool, error) {
+	h.log.TRACE.Printf("read discrete: id %d addr %d qty %d", req.UnitId, req.Addr, req.Quantity)
+	b, err := h.conn.ReadDiscreteInputsWithSlave(req.UnitId, req.Addr, req.Quantity)
+	return h.bytesToBoolResult("read discrete", req.Quantity, b, err)
+}
+
 func (h *handler) HandleCoils(req *mbserver.CoilsRequest) ([]bool, error) {
 	if req.IsWrite {
 		if h.readOnly {
@@ -103,25 +109,25 @@ func (h *handler) HandleCoils(req *mbserver.CoilsRequest) ([]bool, error) {
 		}
 
 		if req.WriteFuncCode == gridx.FuncCodeWriteSingleCoil {
-			h.log.TRACE.Printf("write single coil: id %d addr %d val %t", req.UnitId, req.Addr, req.Args[0])
+			h.log.TRACE.Printf("write coil: id %d addr %d val %t", req.UnitId, req.Addr, req.Args[0])
 			var u uint16
 			if req.Args[0] {
 				u = 0xFF00
 			}
 
 			b, err := h.conn.WriteSingleCoilWithSlave(req.UnitId, req.Addr, u)
-			return h.coilsToResult("write coil", req.Quantity, b, err)
+			return h.bytesToBoolResult("write coil", req.Quantity, b, err)
 		}
 
 		h.log.TRACE.Printf("write coils: id %d addr %d qty %d val %v", req.UnitId, req.Addr, req.Quantity, req.Args)
 		args := coilsToBytes(req.Args)
 		b, err := h.conn.WriteMultipleCoilsWithSlave(req.UnitId, req.Addr, req.Quantity, args)
-		return h.coilsToResult("write coils", req.Quantity, b, err)
+		return h.bytesToBoolResult("write coils", req.Quantity, b, err)
 	}
 
-	h.log.TRACE.Printf("read coil: id %d addr %d qty %d", req.UnitId, req.Addr, req.Quantity)
+	h.log.TRACE.Printf("read coils: id %d addr %d qty %d", req.UnitId, req.Addr, req.Quantity)
 	b, err := h.conn.ReadCoilsWithSlave(req.UnitId, req.Addr, req.Quantity)
-	return h.coilsToResult("read coil", req.Quantity, b, err)
+	return h.bytesToBoolResult("read coils", req.Quantity, b, err)
 }
 
 func (h *handler) HandleInputRegisters(req *mbserver.InputRegistersRequest) ([]uint16, error) {
@@ -142,18 +148,12 @@ func (h *handler) HandleHoldingRegisters(req *mbserver.HoldingRegistersRequest) 
 			return h.exceptionToUint16AndError("write holding", b, err)
 		}
 
-		h.log.TRACE.Printf("write holding: id %d addr %d qty %d val %0x", req.UnitId, req.Addr, req.Quantity, asBytes(req.Args))
+		h.log.TRACE.Printf("write holdings: id %d addr %d qty %d val %0x", req.UnitId, req.Addr, req.Quantity, asBytes(req.Args))
 		b, err := h.conn.WriteMultipleRegistersWithSlave(req.UnitId, req.Addr, req.Quantity, asBytes(req.Args))
 		return h.exceptionToUint16AndError("write multiple holding", b, err)
 	}
 
-	h.log.TRACE.Printf("read holding: id %d addr %d qty %d", req.UnitId, req.Addr, req.Quantity)
+	h.log.TRACE.Printf("read holdings: id %d addr %d qty %d", req.UnitId, req.Addr, req.Quantity)
 	b, err := h.conn.ReadHoldingRegistersWithSlave(req.UnitId, req.Addr, req.Quantity)
 	return h.exceptionToUint16AndError("read holding", b, err)
-}
-
-func (h *handler) HandleDiscreteInputs(req *mbserver.DiscreteInputsRequest) ([]bool, error) {
-	h.log.TRACE.Printf("read discrete: id %d addr %d qty %d", req.UnitId, req.Addr, req.Quantity)
-	b, err := h.conn.ReadDiscreteInputsWithSlave(req.UnitId, req.Addr, req.Quantity)
-	return h.coilsToResult("read discrete", req.Quantity, b, err)
 }

--- a/server/modbus/handler.go
+++ b/server/modbus/handler.go
@@ -14,8 +14,7 @@ import (
 type handler struct {
 	log      *util.Logger
 	readOnly bool
-	mbserver.RequestHandler
-	conn *modbus.Connection
+	conn     *modbus.Connection
 }
 
 func bytesAsUint16(b []byte) []uint16 {
@@ -125,13 +124,13 @@ func (h *handler) HandleCoils(req *mbserver.CoilsRequest) ([]bool, error) {
 	return h.coilsToResult("read coil", req.Quantity, b, err)
 }
 
-func (h *handler) HandleInputRegisters(req *mbserver.InputRegistersRequest) (res []uint16, err error) {
+func (h *handler) HandleInputRegisters(req *mbserver.InputRegistersRequest) ([]uint16, error) {
 	h.log.TRACE.Printf("read input: id %d addr %d qty %d", req.UnitId, req.Addr, req.Quantity)
 	b, err := h.conn.ReadInputRegistersWithSlave(req.UnitId, req.Addr, req.Quantity)
 	return h.exceptionToUint16AndError("read input", b, err)
 }
 
-func (h *handler) HandleHoldingRegisters(req *mbserver.HoldingRegistersRequest) (res []uint16, err error) {
+func (h *handler) HandleHoldingRegisters(req *mbserver.HoldingRegistersRequest) ([]uint16, error) {
 	if req.IsWrite {
 		if h.readOnly {
 			return nil, mbserver.ErrIllegalFunction
@@ -151,4 +150,10 @@ func (h *handler) HandleHoldingRegisters(req *mbserver.HoldingRegistersRequest) 
 	h.log.TRACE.Printf("read holding: id %d addr %d qty %d", req.UnitId, req.Addr, req.Quantity)
 	b, err := h.conn.ReadHoldingRegistersWithSlave(req.UnitId, req.Addr, req.Quantity)
 	return h.exceptionToUint16AndError("read holding", b, err)
+}
+
+func (h *handler) HandleDiscreteInputs(req *mbserver.DiscreteInputsRequest) ([]bool, error) {
+	h.log.TRACE.Printf("read discrete: id %d addr %d qty %d", req.UnitId, req.Addr, req.Quantity)
+	b, err := h.conn.ReadDiscreteInputsWithSlave(req.UnitId, req.Addr, req.Quantity)
+	return h.coilsToResult("read discrete", req.Quantity, b, err)
 }

--- a/server/modbus/proxy.go
+++ b/server/modbus/proxy.go
@@ -22,10 +22,9 @@ func StartProxy(port int, config modbus.Settings, readOnly bool) error {
 	}
 
 	h := &handler{
-		log:            util.NewLogger(fmt.Sprintf("proxy-%d", port)),
-		readOnly:       readOnly,
-		RequestHandler: new(mbserver.DummyHandler), // supplies HandleDiscreteInputs
-		conn:           conn,
+		log:      util.NewLogger(fmt.Sprintf("proxy-%d", port)),
+		readOnly: readOnly,
+		conn:     conn,
 	}
 
 	l, err := net.Listen("tcp", fmt.Sprintf(":%d", port))

--- a/server/modbus/proxy_test.go
+++ b/server/modbus/proxy_test.go
@@ -81,9 +81,8 @@ func TestReadCoils(t *testing.T) {
 	assert.NoError(t, err)
 
 	proxy, _ := mbserver.New(&handler{
-		log:            util.NewLogger("foo"),
-		RequestHandler: new(mbserver.DummyHandler),
-		conn:           downstreamConn,
+		log:  util.NewLogger("foo"),
+		conn: downstreamConn,
 	})
 	assert.NoError(t, proxy.Start(pl))
 	defer func() { _ = proxy.Stop() }()

